### PR TITLE
ARROW-9778: [Rust] [DataFusion] Implement Expr.nullable() and make consistent between logical and physical plans

### PR DESCRIPTION
**Pull request by andygrove**
- Implements `Expr.nullable()`
- Updates some PhysicalExpr nullable methods to be consistent
- Updates `sql.rs` integration test to assert that schema is same between logical and physical plans

### CODE REVIEW

**Comment by andygrove** [comment]
@jorgecarleitao

**Comment by andygrove** [comment]
also @alamb fyi

File: rust/datafusion/src/logicalplan.rs [hunk 1]
```diff
@@ -419,13 +410,50 @@ impl Expr {
...
+            Expr::Alias(expr, _) => expr.nullable(input_schema),
+            Expr::Column(name) => Ok(input_schema.field_with_name(name)?.is_nullable()),
+            Expr::Literal(value) => match value {
+                ScalarValue::Null => Ok(true),
+                _ => Ok(false),
+            },
+            Expr::Cast { expr, .. } => expr.nullable(input_schema),
+            Expr::ScalarFunction { .. } => Ok(true),
+            Expr::AggregateFunction { .. } => Ok(true),
+            Expr::Not(_) => Ok(false),
```
Comment (alamb) [inline rust/datafusion/src/logicalplan.rs:425]:
```suggestion
            Expr::Not(expr) => expr.nullable(input_schema),
```
At least in SQL, `!` can be null, if the input is

File: rust/datafusion/src/logicalplan.rs [hunk 2]
```diff
@@ -419,13 +410,50 @@ impl Expr {
...
+        match self {
+            Expr::Alias(expr, _) => expr.nullable(input_schema),
+            Expr::Column(name) => Ok(input_schema.field_with_name(name)?.is_nullable()),
+            Expr::Literal(value) => match value {
+                ScalarValue::Null => Ok(true),
+                _ => Ok(false),
+            },
+            Expr::Cast { expr, .. } => expr.nullable(input_schema),
+            Expr::ScalarFunction { .. } => Ok(true),
+            Expr::AggregateFunction { .. } => Ok(true),
```
Comment (alamb) [inline rust/datafusion/src/logicalplan.rs:424]: I think the rules here are subtle:
If the aggregate is MIN/MAX/SUM then the nullability is the nullability of the input
If the aggregate is COUNT then the nullability is false (COUNT never can produce NULL)

However, `true` is the safe default, so I think this is ok

File: rust/datafusion/src/logicalplan.rs [hunk 3]
```diff
@@ -419,13 +410,50 @@ impl Expr {
         }
     }
 
+    /// Determine if this expression can produce null values
```
Comment (alamb) [inline rust/datafusion/src/logicalplan.rs:413]:
```suggestion
    /// return true if this expression might produce null values
```

**Review by alamb — APPROVED** [review]
This is a good one 👍  

DataFusion is getting more mature every day :)

**Pull request closed without merging**